add fold_chunks and fold_chunks_with parallel iterators

willcrozi opened the pull request:
This adds two iterators to allow working with fixed-size chunks in a similar way to `chunks()` but without incurring the overhead of allocating a `Vec<T>` for each chunk.

Something similar was discussed in #773.

A call such as:
```
par_iter.fold_chunks(chunk_len, || init(), |acc, x| foo(acc, x))
```
would be functionally equivalent to:
```
par_iter.chunks(chunk_len)
    .map(|chunk|
        chunk.into_iter()
            .fold(init(), |acc, x| foo(acc, x|))
```
A final `fold()` (or `for_each()`, `sum()`, etc) across all mapped chunks can be chained by user code if needed, like this I think it's more flexible.

willcrozi commented:
I've created a small `criterion` benchmark suite to test for various chunk lengths against various equivalent approaches for both 'noop' and 'sum' fold operations: https://github.com/willcrozi/rayon/tree/fold-chunks-bench

Summary graphs here:

![noop_bench](https://user-images.githubusercontent.com/8547876/185607789-57753034-59e6-49f4-a930-ecc96940cab6.svg)
![sum_bench](https://user-images.githubusercontent.com/8547876/185607804-e9c421e9-585f-485b-b165-0723a867e092.svg)

Benchmark runs using the new iterators in this PR have a prefix of `par_foldchunks_...` and `par_foldchunkswith_...` and are coloured yellow and blue respectively.

willcrozi commented:
Hi there! Is there interest in merging something like this? I can see there's some other proposals related to chunking/grouping perhaps making this PR redundant?.

In case there is interest, I feel like there should be a benchmark added in `rayon-demo` since this work is performance related but I don't see an obvious fit in the existing benchmarks. Maybe a new benchmark for some workload where chunking/grouping operations are inherent?

I tried adding a version of the factorial benchmark using `fold_chunks_with` just to see how performance held up. Chunk size was tweaked manually to the best performing power-of-two.
```rust
#[bench]
/// Compute the Factorial using rayon::fold_chunks_with.
fn factorial_fold_chunks_with(b: &mut test::Bencher) {
    fn fact(n: u32) -> BigUint {
        (1..n + 1)
            .into_par_iter()
            .fold_chunks_with(64, BigUint::from(1_u32), |acc, x| acc.mul(x))
            .reduce_with(Mul::mul)
            .unwrap()
    }

    let f = factorial(N);
    b.iter(|| assert_eq!(fact(test::black_box(N)), f));
}
```

Unexpectedly, it's consistently the fastest (ran on Ubuntu 20.04 x86_64 with quad core Sandy Bridge i7-2760QM).
```
running 5 tests
test factorial::factorial_fold_chunks                                   ... bench:   1,396,631 ns/iter (+/- 238,207)
test factorial::factorial_iterator                                      ... bench:  12,819,803 ns/iter (+/- 2,561,952)
test factorial::factorial_join                                          ... bench:   1,590,948 ns/iter (+/- 402,093)
test factorial::factorial_par_iter                                      ... bench:   1,559,018 ns/iter (+/- 314,690)
test factorial::factorial_recursion                                     ... bench:   1,672,888 ns/iter (+/- 399,144)
```

cuviper commented:
Hi! And sorry for not replying sooner. I do think this is compelling, even if we do explore other grouping/chunking ideas as well. This one seems straightforward enough that it can be a relatively easy option to recommend that users try.

Benchmarking is hard, and I'm generally not satisfied with what we have, but your factorial benchmark looks like a good addition, especially since it's winning. :smile: Note that the `join` and `par_iter` versions are converting every number to `BigUint`, causing extra memory allocation overhead, while your fold is multiplying `BigUint * u32`. Here's a closer comparison:

```rust
#[bench]
/// Compute the Factorial using rayon::fold_chunks_with.
fn factorial_fold_with(b: &mut test::Bencher) {
    fn fact(n: u32) -> BigUint {
        (1..n + 1)
            .into_par_iter()
            .fold_with(BigUint::from(1_u32), |acc, x| acc.mul(x))
            .reduce_with(Mul::mul)
            .unwrap()
    }

    let f = factorial(N);
    b.iter(|| assert_eq!(fact(test::black_box(N)), f));
}
```

I get these results on a AMD Ryzen 7 5800X running Fedora 37:

```
running 6 tests
test factorial::factorial_fold_chunks_with                              ... bench:     360,718 ns/iter (+/- 44,535)
test factorial::factorial_fold_with                                     ... bench:     394,420 ns/iter (+/- 25,759)
test factorial::factorial_iterator                                      ... bench:   5,798,113 ns/iter (+/- 30,231)
test factorial::factorial_join                                          ... bench:     414,736 ns/iter (+/- 23,264)
test factorial::factorial_par_iter                                      ... bench:     413,722 ns/iter (+/- 30,028)
test factorial::factorial_recursion                                     ... bench:     697,092 ns/iter (+/- 9,624)
```

However, adding `.with_min_len(64)` to mine does get its result consistently within noise of yours.

Still, that's not to say that we shouldn't add this! I suspect this kind of chunking is probably easier to understand than the loose `with_min_len`/`with_max_len` controls, even though those are more flexible.

adamreichold commented on src/iter/mod.rs:
```diff
@@ -2417,6 +2421,82 @@ pub trait IndexedParallelIterator: ParallelIterator {
+        ID: Fn() -> U + Send + Sync,
+        F: Fn(U, Self::Item) -> U + Send + Sync,
+    {
+        assert!(chunk_size != 0, "chunk_size must not be zero");
```
I think it would be good to mention the panic here and below in the doc comment.

cuviper replied:
And maybe `#[track_caller]` to follow #977.

willcrozi replied:
Thanks for taking a look!

> Note that the `join` and `par_iter` versions are converting every number to `BigUint`, causing extra memory allocation overhead, while your fold is multiplying `BigUint * u32`....
> ...
> However, adding `.with_min_len(64)` to mine does get its result consistently within noise of yours.

Explains my surprise (should have noticed that!)

I'll push my factorial benchmark here shortly. To address the 'like-for-like' issue should it match the `join` and `par_iter` version by mapping everything to `BigUint` first? E.g.
```rust
            .into_par_iter()
            .map(BigUint::from)
            .fold_chunks_with(64, BigUint::from(1_u32), |acc, x| acc.mul(x))
            ...
```
... or maybe supplement with the fastest version your `factorial_fold_with` instead?

willcrozi replied:
> I suspect this kind of chunking is probably easier to understand than the loose `with_min_len`/`with_max_len` controls, even though those are more flexible.

Yes, and also with `fold_chunks` or `chunks` you can guarantee something like: 'all chunks have size _x_ except for the last chunk which may be smaller', a common requirement for protocols/standards. Whereas, if I understand correctly, using `with_min_len`/`with_max_len` controls would not provide such a guarantee(?).

willcrozi replied:
Also, it would be nice to compare against a version using `par_iter().chunks()...` since this was the original motivation for `fold_chunks`: e.g.
```rust
#[bench]
/// Compute the Factorial using rayon::chunks/map and a sequential fold.
fn factorial_chunks_fold(b: &mut test::Bencher) {
    println!("factorial_fold_chunks, chunk size: {}", N);
    fn fact(n: u32) -> BigUint {
        (1..n + 1)
            .into_par_iter()
            .chunks(64)
            .map(|chunk| chunk.into_iter().fold(BigUint::from(1_u32), |acc, x| acc.mul(x)))
            .reduce_with(Mul::mul)
            .unwrap()
    }

    let f = factorial(N);
    b.iter(|| assert_eq!(fact(test::black_box(N)), f));
}
```
However the performance is within noise of `fold_chunks` and the difference seen in the criterion benchmark results above is not seen (use of `BigUint` masks it?). I think it would need to be a whole new benchmark module which didn't require allocation in its 'core' operation, something like summing an iterator of `u32`s into a `u64` perhaps.

willcrozi replied:
Agreed, however there's quite a few places that do the same without mentioning in the doc comment (`IndexedParallelIterator::chunks` and various methods on `ParallelSlice`). Perhaps it would be a cleaner git history to address all of these together in their own PR?

willcrozi commented:
Thanks for taking a look!

> Note that the `join` and `par_iter` versions are converting every number to `BigUint`, causing extra memory allocation overhead, while your fold is multiplying `BigUint * u32`....
> ...
> However, adding `.with_min_len(64)` to mine does get its result consistently within noise of yours.

Explains my surprise (should have noticed that!)

I'll push my factorial benchmark here shortly. To address the 'like-for-like' issue should it match the `join` and `par_iter` version by mapping everything to `BigUint` first? E.g.
```rust
            .into_par_iter()
            .map(BigUint::from)
            .fold_chunks_with(64, BigUint::from(1_u32), |acc, x| acc.mul(x))
            ...
```
... or maybe supplement with the fastest version your `factorial_fold_with` instead?

willcrozi commented:
> I suspect this kind of chunking is probably easier to understand than the loose `with_min_len`/`with_max_len` controls, even though those are more flexible.

Yes, and also with `fold_chunks` or `chunks` you can guarantee something like: 'all chunks have size _x_ except for the last chunk which may be smaller', a common requirement for protocols/standards. Whereas, if I understand correctly, using `with_min_len`/`with_max_len` controls would not provide such a guarantee(?).

willcrozi commented:
Also, it would be nice to compare against a version using `par_iter().chunks()...` since this was the original motivation for `fold_chunks`: e.g.
```rust
#[bench]
/// Compute the Factorial using rayon::chunks/map and a sequential fold.
fn factorial_chunks_fold(b: &mut test::Bencher) {
    println!("factorial_fold_chunks, chunk size: {}", N);
    fn fact(n: u32) -> BigUint {
        (1..n + 1)
            .into_par_iter()
            .chunks(64)
            .map(|chunk| chunk.into_iter().fold(BigUint::from(1_u32), |acc, x| acc.mul(x)))
            .reduce_with(Mul::mul)
            .unwrap()
    }

    let f = factorial(N);
    b.iter(|| assert_eq!(fact(test::black_box(N)), f));
}
```
However the performance is within noise of `fold_chunks` and the difference seen in the criterion benchmark results above is not seen (use of `BigUint` masks it?). I think it would need to be a whole new benchmark module which didn't require allocation in its 'core' operation, something like summing an iterator of `u32`s into a `u64` perhaps.

willcrozi commented:
So I've added my best guess as to what's needed (too many questions probably! :upside_down_face:). I've kept the benchmark performing `BigUint * u32` but also added your version for a fairer comparison.

Let me know if you'd like the commits squashed/rebased onto main etc.

hhirtz commented on src/iter/fold_chunks_with.rs:
```diff
@@ -0,0 +1,286 @@
+        }
+    }
+
+    fn min_len(&self) -> usize {
+        div_round_up(self.base.min_len(), self.chunk_size)
+    }
+
+    fn max_len(&self) -> usize {
+        self.base.max_len() / self.chunk_size
+    }
```
Hi, sorry if it's not an actual issue but if I understand correctly `FoldChunks{,With}Producer::max_len` must also use `div_round_up`, otherwise `min_len` may end up larger than `max_len`. For example with the following configuration:

- `self.base.max_len() == 3`
- `self.base.min_len() == 3`
- `self.chunk_size == 2`

then `min_len == 2` and `max_len == 1`.

willcrozi replied:
I think you're correct, good catch! I took this from the `chunks::ChunkProducer` impl without properly checking :flushed:

@cuviper: shouldn't the impl in `chunks::ChunkProducer` be doing the same thing?

cuviper replied:
Well, you can already set inconsistent values using `with_min_len` and `with_max_len`, and the way it's currently [documented](https://docs.rs/rayon/latest/rayon/iter/trait.IndexedParallelIterator.html#method.with_max_len) and implemented is that the minimum "wins". The maximum is kind of a soft limit, and we might not be able to satisfy both even when the maximum _is_ more than the minimum.

With chunks, either here or the `chunks` adaptor, it seems messier -- but I think it's correct the way it is. We want to find `n` such that `chunk_size * n` respects the inner limit. So for the minimum, we round `n` up to make sure `chunk_size * n` is never less than the inner minimum, and likewise the maximum should round down so `chunk_size * n` is never more than the inner maximum.

With the example of inner `min == max == 3`, and `chunk_size == 2`, there's no possible way to satisfy this entirely. If we pick `n == 1`, then `n * chunk_size == 2` violates the inner minimum, but with `n == 2` then `n * chunk_size == 4` violates the inner maximum. So, I think it's simplest if `min_len` and `max_len` go ahead and compute their ideal limits independently, and then the rest of the code that *uses* these limits will favor the minimum, as I mentioned before.

cuviper commented on src/iter/mod.rs:
````diff
@@ -2417,6 +2421,88 @@ pub trait IndexedParallelIterator: ParallelIterator {
+    /// assert_eq!(chunk_sums, vec![3, 7, 11, 15, 19]);
+    /// ```
+    #[track_caller]
+    fn fold_chunks<ID, F, U>(
````
This is a tiny thing, but maybe this should be `fold_chunks<T, ID, F>` to be consistent with `fold`?

cuviper commented on src/iter/mod.rs:
````diff
@@ -2417,6 +2421,88 @@ pub trait IndexedParallelIterator: ParallelIterator {
+    /// assert_eq!(chunk_sums, vec![3, 7, 11, 15, 19]);
+    /// ```
+    #[track_caller]
+    fn fold_chunks_with<T, F>(
````
But then we have the inconsistency of `fold_with<F, T>`, i.e. `T` second, which I never noticed before...

willcrozi replied:
I've gone with consistency with `fold`, which is consistent with 'std::iter::Iterator::fold` (closure argument goes last).

cuviper reviewed (approved):
Thanks, I think this is good! Just one last consideration on generic params.

I also have some ideas of how to refactor common code between the various "chunks" APIs, but I'll try to follow up separately. That won't affect any public API, at least.

willcrozi commented:
I've noticed there's quite a few tests for `chunks` in `iter/test` and so have added the equivalent to `fold_chunks` and `fold_chunks_with`, I've kept them in their respective modules for now (can move over to `test` module if desired).

There's quite a lot of duplication though so it's understandable if you'd rather not include these.

cuviper commented:
OK, I added another `T: Send` for consistency, and applied test formatting, but otherwise I think I've dithered long enough. :)

bors r+
